### ISSUE TITLE
use/import support

### ISSUE THREAD

**Issue by codeliner** [prolic/fpp#57]
Hi,
first of all, many thanks for this great library. Expect some PR's in near future from me. ;)

I have this in a `types.fpp`

```
namespace App\Model\State {

    data BuildingId = BuildingId deriving (Uuid);

    data User = String deriving (ToString, FromString);

    data Building = Building {BuildingId $buildingId, ?User $user} deriving (ToArray, FromArray);
}

namespace App\Model\Message {

    data AddBuilding = AddBuilding { \App\Model\State\BuildingId $buildingId, \App\Model\State\User $user } deriving (Command: "Building.AddBuilding");

}
```

FPP does an awesome job already. However, it would be nice to import definitions from other namespaces like it is possible in plain PHP or JavaScript (which might be easier to implement and would allow support for multiple FPP files)

## PHP use statement:

```
namespace App\Model\State {

    data BuildingId = BuildingId deriving (Uuid);

    data User = String deriving (ToString, FromString);

    data Building = Building {BuildingId $buildingId, ?User $user} deriving (ToArray, FromArray);
}

namespace App\Model\Message {
    
    use \App\Model\State\BuildingId;
    use \App\Model\State\User;
    //Maybe it is too hard to support something like this: use \App\Model\State;
    //State\BuildingId $building, State\User $user
    //but direct imports would be a nice start

    data AddBuilding = AddBuilding { BuildingId $buildingId, User $user } deriving (Command: "Building.AddBuilding");

}
```

## JS import

JS import statement would have the advantage to specify the fpp file that contains the definition rather than the assigned PHP namespace. 

File `<project_root>/fpp/model.fpp`
```
namespace App\Model\State;

data BuildingId = BuildingId deriving (Uuid);

data User = String deriving (ToString, FromString);

data Building = Building {BuildingId $buildingId, ?User $user} deriving (ToArray, FromArray);
```

File `<project_root>/fpp/messages.fpp`
```
namespace App\Model\Message;

import {BuildingId, User} from '../model';

data AddBuilding = AddBuilding { BuildingId $buildingId, User $user } deriving (Command: "Building.AddBuilding");
```
Note: Import path is relative to current file (or maybe absolute to project root?). File extension `.fpp` is not needed. FPP would need to detect conflicting namespaces.

**Comment by prolic** [comment]
> first of all, many thanks for this great library. Expect some PR's in near future from me. ;)

you kidding me? :+1:

**Comment by codeliner** [comment]
> you kidding me?

just a lovely nostal insider, saying exactly what I wanted to say ;)

**Comment by codeliner** [comment]
@prolic any preferences?

My favorite is JS import because supporting multiple FPP files helps to keep things organized.

**Comment by prolic** [comment]
I'm not sure, using PHP imports has the benefit, that you can refactor easily and move stuff around, while with JS import you have to change all definitions, if you move a FPP definition to another file.
Also:  `import {BuildingId, User} from '../model';` - what if two BuildingId's are stored within the model.fpp file with different namespaces, which one to chose?

**Comment by codeliner** [comment]
> I'm not sure, using PHP imports has the benefit, that you can refactor easily and move stuff around,

Ok, this could work but FPP needs to scan all files on each run to know which namespaces and definitions are available.

PHP namespaces are mapped to a folder structure but FPP is not (at the moment). So FPP cannot locate the file that contains a definition imported by a use statement.

Also I see FPP files more like JS ES6 module files and not like typical PHP files (containing one class per file), because you'll likely group several fpp definitions - but not all - in one file.

> what if two BuildingId's are stored within the model.fpp file with different namespaces, which one to chose?

Good point! I'd say in that case you have to import using the FQCN or even one step further import with an alias:

`import {\App\Model\State\BuildingId as StateBuilidngId, \Ap\API\BuildingId as ApiBuildingId} from '../model'`;

But, I'd say in a naive implementation FPP could just throw some kind of conflict exception and force the developer to use a simpler structure.

**Comment by prolic** [comment]
Another argument against JS imports: FPP doesn't track from which file it read the definitions. After parsing all definitions are stored in the DefinitionCollection, so matter where they come from.

**Comment by codeliner** [comment]
I'm using [this config](https://github.com/prolic/fpp/issues/56) to let PHPStorm run FPP whenever I change a FPP file: (like documented in the wiki but using docker)

In that case FPP only handles current file. Ok, I could modify the command so that FPP scans the entire project dir over and over again no matter what file has changed. Doing this would ensure that all definitions are in the DefinitionCollection. But that's a performance nightmare in real world projects.

## Options
1.  Long-running fpp process watching file changes itself and keeping information of all scanned files in memory. This solution would work well with PHP-like use statements, because we know all available definitions and their namespaces.

2. Start file scan by resolving imports and do that recursively until all dependent files are scanned and DefinitionCollection is filled up. Then process current file.

Option 2 sounds easier. It should be easier to develop and I think it's also easier to use. But how can option 2 be combined with PHP-like use statements? They don't tell the parser where to find related fpp files. We would need to define a PSR-like convention, but conventions introduce other problems.

Refactoring is an issue. In the JS ES6 world modern IDEs like WEBStorm or Visual Studio Code refactor imports automatically when files are moved around. Sooner or later a PHPStorm fpp plugin is required anyway. Maybe we could solve refactoring issues with proper PHPStorm integration?

**Comment by prolic** [comment]
I think option 2 is good (and we can use JS imports then). Wanna provide a PR?

I wouldn't bother about even better PHPStorm integration (incl. refactoring feature) too much for now. I don't know how to write those plugins and the company probably won't provide a plugin for it by themselves any time soon.

**Comment by codeliner** [comment]
> I think option 2 is good (and we can use JS imports then). Wanna provide a PR?

Good news. Yes, I'd like to tackle the issue. But need a time slot for it. Maybe I can start working on it next week.

**Comment by prolic** [comment]
How does it work with fpp.lock file for deletion of removed objects? Seems
like a problem to me, when we don't parse all the fpp files.

**Comment by codeliner** [comment]
fpp.lock could contain a map of `File => OutputObject[]`

When a file is changed, fpp checks the map and compares is with current objects defined in the file.

The only real problem is how we can detect deletion of an entire fpp file? 
But that's a problem of the full scan approach as well, because I think that PHPStorm does not trigger the file watcher in case a file is deleted.
So even in that scenario deleted fpp files are first recognized when another fpp file is modified.

A workaround would be to first empty the file, hit save to trigger file watcher, let fpp clean up the objects and finally delete the empty file. This along with a `fpp remove $file` command is maybe enough.

A long-running fpp process using file system file watchers would probably be the better option, but that's too mich work for current level of the project.

**Comment by prolic** [comment]
Maybe this can help?
https://github.com/jasonlewis/resource-watcher/blob/master/README.md

**Comment by codeliner** [comment]
Not sure about performance impact. Readme says that it uses a polling strategy to detect changes. That might work well for a single directory but recursively scanning project `src` every second? Could become a problem in larger projects.

A convention that all fpp files should be located somewhere outside of `src` could work but as I said earlier: IMHO such a convention is a bad idea. My structuring plan (that brought me to this import topic) is to keep fpp files next to generated php files
```
|_ Command
  |_ _commands.fpp
  |_ RegisterUser.php
  |_ ActivateUser.php
|_ Event
  |_ _events.fpp
  |_ UserRegistered.php
  |_ UserActivated.php
|_ User
  |_ _state.fpp
  |_ UserData.php
  |_ UserId.php
  |_ ActivationLink.php  
```

I'm sure other users will have different preferences of how they want to organize fpp and generated php files. It is a very interesting question, anyway. Time will show if a best practice can be defined.

**Comment by prolic** [comment]
https://github.com/emcrisostomo/fswatch/blob/master/README.md then?

**Comment by codeliner** [comment]
damn, there is another issue:

```
namespace App\Model\Message;

import {BuildingId, User} from '../model';

data AddBuilding = AddBuilding { BuildingId $buildingId, User $user } deriving (Command: "Building.AddBuilding");
```

If `BuildingId` is changed or removed, fpp needs to parse `messages.fpp` again. `fpp.lock` can contain a second mapping, though.

**Comment by prolic** [comment]
Maybe it's still best to parse all fpp files at once (hence php imports,
not js). I would recommend to collect them in an fpp dir (outside src), to
avoid performance problems. Solves a lot of issues.

**Comment by codeliner** [comment]
looks like a really good lib @prolic 

But IMHO [installation](https://github.com/emcrisostomo/fswatch/blob/master/README.md#installation) is a problem. At the moment fpp requires nothing but standard PHP and maybe setting up PHPStorm for smooth development integration.

Using fswatch would require a complex installation (depending on OS) and that's a killer for adoption.
For me it wouldn't be a problem. I'd create a docker image but standard PHP only is a very strong argument against fswatch.

**Comment by prolic** [comment]
Maybe we can copy the binaries needed, so everything comes shipped at once.

Also if all fpp files are together (outside src) then polling ain't such a
big issue.

**Comment by codeliner** [comment]
OK, what about this? :

We go for PHP use statements and scan all files on each run (If this gets to slow for large projects we might optimize it later with clever indexing and change detection).

FPP keeps a list of scanned files in `fpp.lock`.

We have two ways to invoke FPP:

## Invoke with dir path

FPP compares all files in the dir with the list of known fpp files in `fpp.lock` of same directory. If a file is no longer present in the directory FPP takes information from `fpp.lock` to delete generated PHP files connected with missing fpp file (index maintained in fpp.lock).

## Invoke with file path  

FPP checks if file is known in `fpp.lock`. If not the file is added to that list.

## Scanning Phase

Once file comparison is finished FPP iterates over list of known files in `fpp.lock` to scan all.
This way it does not matter if all files are located in the same directory or spread across project `src`.
If files are in different locations this simply means that FPP cherry-picks the files without the need to scan all directories under `src`

Putting all fpp files in one directory would have the advantage that file deletions can be detected on later runs. But if one can live with the trade off that an fpp needs to be emptied before final deletion than they can use my suggested (and preferred) structure as well.

If we can agree on that I'd add PHP use statement support in a PR and `fpp.lock` along with the behavior described above can be added in another PR.

**Comment by prolic** [comment]
PHPStorm comes already with a file watcher. Have you tried putting the fpp files into src dir somewhere? If so, is it really a performance problem to scan the src-dir every time? Maybe that's not a real problem at all (at least with a SSD).

**Comment by codeliner** [comment]
What's the problem with my suggestion? There is no need to scan the entire project directory if fpp maintains a cache like described in my last comment. No convention for file location needed. So I thought everything is fine now. FPP can of course be invoked using PHPStorm's file watcher. I don't suggest anything different here.

**Comment by prolic** [comment]
Sounds good, go for it.

I was just curious if performance of scanning src dir for fpp files is really bad or if this was just an assumption.

**Comment by codeliner** [comment]
ah ok. Yeah, it is an assumption. I just played around with FPP so project is very small.

What I can tell is that the [message flow analyzer](https://github.com/prooph/message-flow-analyzer) needs some time to recursively scan a large project. But it touches every file and parses nearly every class. So a lot of performance is lost in processing logic. 
Another observation is TypeScript + webpack running in watch mode. Sometimes this makes heavy use of system resources.

The good thing is, supporting PHP use statements and indexing fpp files are two separate topics. We can start with use statements and just test this by simulating a larger project. Maybe performance is good enough. And if not, we can add the indexing strategy.

**Comment by prolic** [comment]
As a short test I scanned my complete home-dir on the notebook. About 58GB of data, took 13 secs to scan.

**Comment by simensen** [comment]
@codeliner Are you still working on this?

I'm splitting off part of my experience reported in #115 to here as it looks like part of what I'd like to discuss seems to have been discussed here.

Not being able to alias or `use` class names means that even within the same `.fpp` file things get quite hard to read. For a real world example, the following is one of the files I created last night and it is a bear to read due to the repetition of the namespace prefixes:

```
namespace App\Model\User {
    data UserId = UserId deriving (Uuid);
    data Name = String deriving (ToString, FromString);
    data Username = String deriving (ToString, FromString);
    data EmailAddress = String deriving (ToString, FromString);
    data HashedPassword = String deriving (ToString, FromString);
    data PlaintextPassword = String deriving (ToString, FromString);
}

namespace App\Model\User\Command {
    data CreateUser = CreateUser {
        \App\Model\User\UserId $userId,
        \App\Model\User\Name $name,
        \App\Model\User\Username $username,
        \App\Model\User\EmailAddress $emailAddress,
        \App\Model\User\HashedPassword $hashedPassword
    } deriving (Command);

    data RenameUser = RenameUser {
        \App\Model\User\UserId $userId,
        \App\Model\User\Name $name
    } deriving (Command);

    data ChangeUsername = ChangeUsername {
        \App\Model\User\UserId $userId,
        \App\Model\User\Username $username
    } deriving (Command);

    data ChangePassword = ChangePassword {
        \App\Model\User\UserId $userId,
        \App\Model\User\PlaintextPassword $plaintextPassword
    } deriving (Command);
}

namespace App\Model\User\Event {
    data UserCreated = UserCreated {
        \App\Model\User\UserId $userId,
        \App\Model\User\Name $name,
        \App\Model\User\Username $username,
        \App\Model\User\EmailAddress $emailAddress,
        \App\Model\User\HashedPassword $hashedPassword
    } deriving (AggregateChanged);

    data UserRenamed = UserRenamed {
        \App\Model\User\UserId $userId,
        \App\Model\User\Name $name
    } deriving (AggregateChanged);

    data UsernameChanged = UsernameChanged {
        \App\Model\User\UserId $userId,
        \App\Model\User\Username $username
    } deriving (AggregateChanged);

    data PasswordChanged = PasswordChanged {
        \App\Model\User\UserId $userId,
        \App\Model\User\HashedPassword $hashedPassword
    } deriving (AggregateChanged);
}
```

It seems like we should be able to make this easier for the 95% cases more quickly if we don't worry about referencing other files or other classes. As noted in #115 I can see that is a potential problem for me eventually... but the more immediate problem I see is not being able to read, write and maintain `.fpp` files due to the excessive repetition of the namespace prefix.

If this proposal looks interesting and @codeliner doesn't have a solution yet, does my proposal sound interesting? It is 

```
namespace App\Model\User {
    data UserId = UserId deriving (Uuid);
    data Name = String deriving (ToString, FromString);
    data Username = String deriving (ToString, FromString);
    data EmailAddress = String deriving (ToString, FromString);
    data HashedPassword = String deriving (ToString, FromString);
    data PlaintextPassword = String deriving (ToString, FromString);
}

//
// Importing fully-qualified class names and aliases from a namespace.
//
// Implies that fpp itself knows about these namespaces already. In
// this case, they are defined earlier in the file.
//

namespace App\Model\User\Command {
    // option 1
    use {UserId, Name, Username, EmailAddress, HashedPassword as Password} from \App\Model\User;

    // option 2
    use \App\Model\User\UserId;
    use \App\Model\User\Name;
    use \App\Model\User\Username;
    use \App\Model\User\EmailAddress;
    use \App\Model\User\PlaintextPassword as Password;

    data CreateUser = CreateUser {
        UserId $userId,
        Name $name,
        Username $username,
        EmailAddress $emailAddress,
        Password $password
    } deriving (Command);

    data RenameUser = RenameUser {
        UserId $userId,
        Name $name
    } deriving (Command);

    data ChangeUsername = ChangeUsername {
        UserId $userId,
        Username $username
    } deriving (Command);

    data ChangePassword = ChangePassword {
        UserId $userId,
        Password $password
    } deriving (Command);
}

//
// Importing fully-qualified namespace prefixes.
//
// Implies that fpp itself knows about these namespaces already. In
// this case, they are defined earlier in the file.
//

namespace App\Model\User\Event {
    use \App\Model\User;

    data UserCreated = UserCreated {
        User\UserId $userId,
        User\Name $name,
        User\Username $username,
        User\EmailAddress $emailAddress,
        User\HashedPassword $hashedPassword
    } deriving (AggregateChanged);

    data UserRenamed = UserRenamed {
        User\UserId $userId,
        User\Name $name
    } deriving (AggregateChanged);

    data UsernameChanged = UsernameChanged {
        User\UserId $userId,
        User\Username $username
    } deriving (AggregateChanged);

    data PasswordChanged = PasswordChanged {
        User\UserId $userId,
        User\HashedPassword $hashedPassword
    } deriving (AggregateChanged);
}
```

Or we could use php7 style `use` if that is preferred:

```
use \App\Model\User\{UserId, Name, Username, EmailAddress, HashedPassword as Password};
```

I'd probably also implement it such that `use` statements would not require the leading `\` but that is something I'm not 100% sure I care too much about:

```
use App\Model\User\{UserId, Name, Username, EmailAddress, HashedPassword as Password};
```

**Comment by prolic** [comment]
@simensen I'm acutally okay with any solution, as long as it's not adding too much boilerplate again (the reason this lib exists is to reduce boilerplate). If you want to work on this to improve, you're welcome, I currently don't have enough time and event-store-client has higher priority to me at the moment.

**Comment by codeliner** [comment]
@simensen never found the time to work on this. TBH FPP is a great idea but it is too difficult to customize. We are using value objects quite heavily and let them even handle parts of the business logic. Hence, I've searches for a way to generate boilerplate code without leaving PHP and my solution is a mix of PHPstorm templates and a trait for handling complex types. It's documented here: https://event-engine.io/api/immutable_state.html

**Comment by prolic** [comment]
A release 0.1.0 has been tagged. There is currently a rewrite going on, so I will close all open issues. If you still want to resolve this issue yourself, you can submit a pull request to the 0.1.x branch.
